### ISSUE TITLE
Easier way to create tensors with names

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#27175]
## 🚀 Feature

Right now, to create a tensor with names, we do
```
x = torch.randn(3, 3, 3, 3, names=('N', 'C', 'H', 'W'))
```

Typing the ' a lot is an inconvenience. It would be nice to explore one of the following three APIs:

1.  `x = torch.randn(N=3, C=3, H=3, W=3)`
2. `x = torch.randn(torch.dims(N=3, C=3, H=3, W=3))`
3. `x = torch.randn(3, 3, 3, 3, names='N C H W')
```

## Motivation

Type less, feel happier.

**Comment by vadimkantorov** [comment]
`names = 'NCHW'` (without spaces) should probably work out-of-the-box? if names are accepted as a generic sequence

**Comment by andreaskoepf** [comment]
Can named tensors still be indexed 'classically' without specifying dimension names? If yes this would require ordered `**kwargs` for the proposals 1. and 2., e.g. Python >= 3.6 (see [PEP 468](https://www.python.org/dev/peps/pep-0468/)). Another reason to drop Python 2 support https://github.com/pytorch/pytorch/issues/23795.

**Comment by zou3519** [comment]
> Can named tensors still be indexed 'classically' without specifying dimension names? If yes this would require ordered `**kwargs` for the proposals 1. and 2., e.g. Python >= 3.6 (see [PEP 468](https://www.python.org/dev/peps/pep-0468/)). Another reason to drop Python 2 support #23795.

Yes. I neglected to mention the ordering constraint for Python >= 3.6 thank you for pointing it out.

**Comment by alok** [comment]
@vadimkantorov Wouldn't that limit each name to a single character since parsing would quickly become ambiguous? (imagine `(time)nchw` for 5 dimensions where the parens are for clarity).

**Comment by vadimkantorov** [comment]
It would, but if one wants longer names, they can still use the full tuple syntax

**Comment by michalwols** [comment]
A comma separated string might be cleaner since it would allow skipping dimensions

```python
x = torch.randn(3, 3, 3, 3, 'N,,H,W')
```

and multiple character names
```python
x = torch.randn(32, 3, 3, 224, 224, 'time,n,c,h,w')
```

and potentially glob like syntax to drop rest (similar to list/tuple unpacking)
```python
x = torch.randn(32, 3, 3, 224, 224, 'time,n,*')
```

**Comment by michalwols** [comment]
@zou3519 another thing that would be nice is having the torch.Size() tuple be a namedtuple for named tensors so that you could do things like

```python
embs = flatten(feats)
assert feats.shape.height * feats.shape.width == embs.shape.dim
```

to perform shape validation. Inspired by the new static shape validation work that @apaszke did for TF Swift.

**Comment by alok** [comment]
I do like the idea of making shape a (subclass of?) namedtuple.

--- Alok

**Comment by michalwols** [comment]
yeah would probably need to be a subclass of tuple to support missing names

**Comment by ekshaks** [comment]
We have a couple of options in the [tsalib](https://github.com/ofnote/tsalib) library. Please see the notes on the [shorthand](https://github.com/ofnote/tsalib/blob/master/notebooks/shorthand.md) notation. In short:

- declare all named dimensions upfront.
- write 'btd' (dim names without separators) when each dim name is a single character. Otherwise, use comma separators 'b,t,dim'. This is almost the same effort as 'b t d'.
- allow specifying anonymous dimensions using '_', e.g., 'b_d'.

While tsalib allows the full tuple `(B,T,D)` style specs, we've found it to be too verbose. Instead, it is less tiring to simply declare the named dimensions with their shorthands ('B(b):10' or just 'B(b)') upfront, and then use the shorthands `b`, `t`, etc, everywhere. 

Hope this helps.

**Comment by ekshaks** [comment]
I think the shorthand shape transform specification, as we use in [tsalib](https://github.com/ofnote/tsalib) may also be useful to make code concise with namedtensors.

Examples:
```python
# instead of 
flat_imgs = imgs.flatten(['C', 'H', 'W'], 'features')
# write
flag_imgs = imgs.flatten('CHW -> features')
```
```python
# instead of 
 named_tensor.align_to('F', 'E', ...)
# write
named_tensor.align_to('FE*')
```

I think one could push it further (too far? :) and simply use the `warp` operator, where multiple shape transformations can be chained together with `->`.
